[Discover] Drag & drop for adding columns to the table

jughosta opened the pull request:
Closes https://github.com/elastic/kibana/issues/151703

## Summary

This PR:
- changes design of field list item in Lens and Discover
- changes design of dimension triggers in Lens panels
- changes design "Add layers" button
- adds "+" action to field items in Lens
- unifies the logic of rendering Lens field items for text-based and form-based views
- adds Drag&Drop functionality to Discover to add columns to the table by dragging a field from the sidebar
- adds functional tests for drag&drop in Discover
- shows fields popover in text-based mode too so users can copy field names

[Figma](https://www.figma.com/file/SvpfCqaZPb2iAYnPtd0Gnr/KUI-Library?node-id=674%3A198901&t=OnQH2EQ4fdBjsRLp-0)

<details>
<summary>Gifs</summary>

![Apr-04-2023 14-38-47](https://user-images.githubusercontent.com/1415710/229795117-712267ba-f5e0-42ca-a2e5-e23759d5ddda.gif)
![Apr-04-2023 14-40-59](https://user-images.githubusercontent.com/1415710/229795133-7b618566-e73a-4303-97d7-b2840d1fc006.gif)

</details>


### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

andreadelrio reviewed (changes requested):
@jughosta this is looking great! I noticed that after grabbing a field it is as though the item is stuck something like hover state (hence the quick add/remove button is present). See this video, field `extension`. The same happens when just grabbing and letting go of a field, even if it isn't added to the table.

https://user-images.githubusercontent.com/4016496/229948471-f0129523-9026-406b-9ac4-772410b0ca8a.mov

-- 

I was thinking we could improve the mixin `mixinDomDroppable` to make the drop area stand out a bit more:

` border: $euiBorderWidthThin dashed $euiBorderColor !important;` to
` border: $euiBorderWidthThick dashed $euiColorVis0 !important;`

Here's the current vs proposed mixin in screenshots:

<img width="1079" alt="Frame 3" src="https://user-images.githubusercontent.com/4016496/229954222-b78da032-a1b8-4e10-86c3-9b54a78a3d01.png">

walterra reviewed (approved):
ML changes LGTM.

jughosta commented:
@andreadelrio Thanks a lot for the detailed feedback and suggestions!

- I pushed a small change which should improve the "stuck" hover state
- Made the dashed border thicker for "green" drop zones as suggested. In Lens there are also extra drop zone popovers ("Duplicate", "Swap" etc) which I also had to update to avoid vertical jumps. Does it look fine?

![Apr-05-2023 12-36-38](https://user-images.githubusercontent.com/1415710/230057580-6e2d0a13-d8e6-477d-824c-71ca74ecc90b.gif)

davismcphee commented:
Added @1Copenut as a reviewer for a11y feedback.

andreadelrio commented:
> @andreadelrio Thanks a lot for the detailed feedback and suggestions!
> 
> * I pushed a small change which should improve the "stuck" hover state
> * Made the dashed border thicker for "green" drop zones as suggested. In Lens there are also extra drop zone popovers ("Duplicate", "Swap" etc) which I also had to update to avoid vertical jumps. Does it look fine?
> 
> ![Apr-05-2023 12-36-38](https://user-images.githubusercontent.com/1415710/230057580-6e2d0a13-d8e6-477d-824c-71ca74ecc90b.gif) [ ![Apr-05-2023 12-36-38](https://user-images.githubusercontent.com/1415710/230057580-6e2d0a13-d8e6-477d-824c-71ca74ecc90b.gif) ](https://user-images.githubusercontent.com/1415710/230057580-6e2d0a13-d8e6-477d-824c-71ca74ecc90b.gif) [ ](https://user-images.githubusercontent.com/1415710/230057580-6e2d0a13-d8e6-477d-824c-71ca74ecc90b.gif)

@jughosta Yes, it looks good to me. Thanks for making the changes.

1Copenut commented on test/functional/page_objects/discover_page.ts:
```diff
@@ -868,4 +868,51 @@ export class DiscoverPageObject extends FtrService {
+  }
+
+  /**
+   * Drags field with keyboard actions to add as a column
```
Glad to see this added to the test suite!

jughosta replied:
@1Copenut Thank you very much for a such detailed feedback! I created a separate issue for a follow up work https://github.com/elastic/kibana/issues/155015

1Copenut reviewed (approved):
I reviewed the code by file looking for any obvious a11y challenges or items for improvement. Seeing none, I road tested the Discover drag and drop using Safari, Firefox, and Chrome on Mac, and Chrome on Windows. Did the testing listed below. Notes to follow after.

I really appreciate the chance to test this design as you're working on it. This is a big help for establishing good screen reader and keyboard UX, and it's fun to try out new features.

Let me know if you have any questions about testing or notes/feedback below.

**MacOS**
* Keyboard navigation - ensured I could drag, drop, cancel, reorder the table
* Zoomed layout up to 200% at 1920px width - navigated with keyboard
* Dragged and dropped using my mouse
* Tested the drag and drop interface with Safari + VoiceOver. Worked very well.

**Windows 10**
* Keyboard navigation - ensured I could drag, drop, cancel, reorder the table
* Zoomed layout up to 200% at 1920px width - navigated with keyboard
* Tested the drag and drop interface with Chrome + NVDA. Worked well after I figured out I needed to switch into [Focus mode](https://dequeuniversity.com/screenreaders/nvda-keyboard-shortcuts#nvda-browse_focus_modes).

**Notes**
* The interface allows me to do all the actions a user might undertake with a mouse. I did not detect any accessibility issues that gave me pause.
* If there is not an `axe-core` scan check in the test suite, I'd recommend adding a test after dragging a few things into the data grid.
* The controls could use some screen reader only language about how to use them. I almost thought NVDA would not work with the DND interface, but tried out Focus mode on a whim. We should strongly consider adding instructions as an `EuiScreenReaderOnly` component with an ID that can be referenced as an `aria-describedby` attribute on the Available, Common, and Selected accordion buttons. This gives users several changes to hear how to work the controls.
* I'm still looking for a skip link between sections of drag and drop controls. There's a lot of tabbing that has to happen if I'm keyboard only.
* There are 3 buttons in each control, which is traversable but verbose for screen readers and keyboard. The skip link would alleviate some of this. The three buttons are tougher, because one's to drag, one's to open the popover, one's to delete the selected item from your data grid.
* Is there a way to remove selected items from the Available Items group? If I add an input, does it make sense to allow me to grab it to add again? Small UX nit, but worth considering.

andreadelrio reviewed (approved):
Design changes LGTM. Great job 🚀

andreadelrio commented on packages/kbn-dom-drag-drop/src/sass/drag_drop.scss:
```diff
@@ -1,22 +1,21 @@
   left: 0;
   opacity: .9;
-  transform: translate(-12px, 8px);
+  transform: translate(4px, 4px);
```
@jughosta please use EUI variables here.

kertal commented on src/plugins/discover/public/application/main/components/sidebar/discover_field.tsx:
```diff
@@ -8,162 +8,70 @@
+} {
+  const dataTestSubj = `fieldToggle-${field.name}`;
+  const handler =
+    field.name === '_source' ? undefined : (f: DataViewField) => toggleDisplay(f, isSelected);
```
Just a note, and nothin actionable, I think we can be sure that the `_source` field won't be part of the field list given we use the field_caps API ... but out of scope of this PR

kertal commented on src/plugins/unified_field_list/public/components/field_item_button/field_item_button.tsx:
```diff
@@ -22,72 +22,211 @@ import './field_item_button.scss';
+  const fieldAction = isSelected
+    ? onRemoveFieldFromWorkspace && (
+        <EuiToolTip
+          key={`selected-to-remove-${field.name}-${removeFieldFromWorkspaceTooltip}`}
```
Question, nothing blocking, is such a long `key` necessary? Is there even a key necessary?

jughosta replied:
@kertal Yes, the tooltip component needs to be recreated otherwise it was appearing incorrectly after pressing a button.

kertal commented on src/plugins/unified_field_list/public/components/field_item_button/field_item_button.tsx:
```diff
@@ -22,72 +22,211 @@ import './field_item_button.scss';
+      )
+    : onAddFieldToWorkspace && (
+        <EuiToolTip
+          key={`deselected-to-add-${field.name}-${addFieldToWorkspaceTooltip}`}
```
same like the previous comment

stratoula commented:
There are some significant changes on the Lens dimension editor

<img width="1741" alt="image" src="https://user-images.githubusercontent.com/17003240/232701113-954a6b1a-dfc5-4b38-bc23-87e85b4335aa.png">


Also drag and drop styling has changed

@MichaelMarcialis can you take a look first and if this is ok by you then I will continue my review

MichaelMarcialis commented:
Thanks so much for putting this together, @jughosta! And thank you for taking the design lead on this @andreadelrio! This is wonderful. I gave this a test drive in Lens and left a few quick comments below. Let me know if you have any questions.

- Would it be possible to reduce the height of the `EuiColorPaletteDisplay` components that conditionally appear in Lens dimension items to 2px (rather than the current height of 4px)? I realize that would be a CSS override to the EUI component, but I think it will look better in conjunction with the newly reduced dimension item height.
  
  ![image](https://user-images.githubusercontent.com/3884767/232805981-6161daf7-2d4d-46a3-b5e9-61b83e9fbca7.png)

- Can we change the dimension drop zone icon from `plusInCircleFilled` to a regular `plus`?
  
  ![image](https://user-images.githubusercontent.com/3884767/232806296-d577da6b-dfad-4da3-87de-118b6ee7a8ac.png)

- Can we also change the dimension drop zone icon and text color to `euiColorDisabledText` (in its default state, when nothing is being dragged)?
  
  ![image](https://user-images.githubusercontent.com/3884767/232806474-e2112365-d92d-49ff-9d90-3135bf8c0b18.png)

- Is it possible to ensure the button element that triggers the configuration flyout inside each dimension item extends to the full available width of that dimension item (similar to the field list items)? Currently, if the dimension item has a short name, the right half of the dimension item doesn't open the flyout when clicked.
  
  ![image](https://user-images.githubusercontent.com/3884767/232806903-0ea50827-19f4-40c0-8207-ac4f0ccfd37d.png)

- I know I said I'd be OK with increasing the border width of all drop zones to 2px in a discussion with @andreadelrio. However, after seeing it in practice, the border width increase feels a bit out-of-place. Was there a compelling reason to have the drop zones use a 2px border instead of the previous 1px? If not, would ya'll be opposed to bringing it back to 1px? Sorry to flip-flop on this issue!

andreadelrio commented:
> * I know I said I'd be OK with increasing the border width of all drop zones to 2px in a discussion with @andreadelrio. However, after seeing it in practice, the border width increase feels a bit out-of-place. Was there a compelling reason to have the drop zones use a 2px border instead of the previous 1px? If not, would ya'll be opposed to bringing it back to 1px? Sorry to flip-flop on this issue!

@MichaelMarcialis The reason is that with 2px the dropzone doesn't stand out as much in Discover because the document table is there as opposed to the prominent (with white background) empty prompt that Lens has. See:

<img width="1079" alt="Frame 3" src="https://user-images.githubusercontent.com/4016496/232860596-1286f847-de89-4bbf-809b-f6b11d0ab0f5.png">

@jughosta my recommendation would be to revert the changes in `mixinDomDroppable`. We can either create a separate mixin (e.g. `mixinDomDroppableThick` or pass a width variable (`borderWidth` or `width`) to the original `mixinDomDroppable`. That way we don't affect Lens.

jughosta commented:
@MichaelMarcialis @andreadelrio 

Pushed new updates. Could please review again? Thanks!

> Would it be possible to reduce the height of the EuiColorPaletteDisplay components that conditionally appear in Lens dimension items to 2px

Done

> Can we change the dimension drop zone icon from plusInCircleFilled to a regular plus?

Done

> Can we also change the dimension drop zone icon and text color to euiColorDisabledText (in its default state, when nothing is being dragged)?

The button is clickable in the default state - it's not disabled. Would it be okay to rather keep the current color?

> Is it possible to ensure the button element that triggers the configuration flyout inside each dimension item extends to the full available width of that dimension item (similar to the field list items)? 

Done

>  If not, would ya'll be opposed to bringing it back to 1px?

> my recommendation would be to revert the changes in mixinDomDroppable. We can either create a separate mixin (e.g. mixinDomDroppableThick) or pass a width variable (borderWidth or width) to the original mixinDomDroppable. That way we don't affect Lens.

Done. For Lens it's back to 1px border width. For Discover it's still 2px.

![lens](https://user-images.githubusercontent.com/1415710/233030277-10e02b20-f2d5-4d05-bfd7-ac43ca5f7b81.gif)
![discover](https://user-images.githubusercontent.com/1415710/233030296-fc4efa3e-250f-4f07-9a5c-b060bcb921d7.gif)

stratoula reviewed (commented):
Some comments from my review

- For some reason all the fields in Lens text based languages mode look like text fields
<img width="867" alt="image" src="https://user-images.githubusercontent.com/17003240/233047884-eba5965b-f717-40c5-bb1e-2cb30304e67d.png">

This works fine in main

- When there is a warning on the breakdown I see 2 icons
<img width="426" alt="image" src="https://user-images.githubusercontent.com/17003240/233048725-b2fe10ec-4225-4711-9e00-89fab6acddc1.png">

but with this change the palette icon has been vanished
<img width="414" alt="image" src="https://user-images.githubusercontent.com/17003240/233048799-c2ec9e38-0991-4531-880b-c060ff998f47.png">

Is this intended?

stratoula commented:
Also in text based languages we dont want to show the field popover

<img width="634" alt="image" src="https://user-images.githubusercontent.com/17003240/233052614-a9d71841-1562-4a90-9625-d87ef9d14749.png">

Can we remove it (it doesnt add any value to show an empty popover with just the field title and the + icon

jughosta commented:
@stratoula 

Good catch, thanks a lot!

> For some reason all the fields in Lens text based languages mode look like text fields

Fixed

> but with this change the palette icon has been vanished

Fixed. It was necessary to accommodate some space for a drag handle but it was decided in the end to drop the left handle.

> Can we remove it (it doesnt add any value to show an empty popover with just the field title and the + icon

We can remove it but this is the only way how users can now copy field names. Would it be okay to keep it?

MichaelMarcialis reviewed (approved):
Left two small comments below, but this looks great to me. Approving now so I don't hold you up. Thanks, @jughosta!

- The dimension item name text is user-selectable, while the text in the field list items is not. Should we set the dimension item text to `user-select: none;` for consistency (and to prevent it from being highlighted when clicked, as shown below)?
  
  ![image](https://user-images.githubusercontent.com/3884767/233098902-7da60d5c-ab52-41d1-a55e-a47db7966b34.png)

MichaelMarcialis commented on packages/kbn-dom-drag-drop/src/sass/drag_drop.scss:
```diff
@@ -91,12 +94,12 @@
 }
 
-$lnsLayerPanelDimensionMargin: 8px;
+$reorderItemMargin: $euiSizeS;
```
```suggestion
```

MichaelMarcialis commented on packages/kbn-dom-drag-drop/src/sass/drag_drop.scss:
```diff
@@ -91,12 +94,12 @@
   width: 100%;
   top: 0;
-  height: calc(100% + #{$lnsLayerPanelDimensionMargin});
+  height: calc(100% + #{$reorderItemMargin / 2});
```
```suggestion
  height: calc(100% + #{$euiSizeS / 2});
```

jughosta replied:
Thanks for the suggestion, @MichaelMarcialis! 
Having a named variable was helpful during this redesign work: easier to find what to needs to change if we change size of items. I would like to keep the variable.

MichaelMarcialis reviewed (approved):
@jughosta: Thanks for those updates! Final nitpick suggestion below, but this is good to go from my perspective.

As for the `user-select` thing, you're right. I mistook the focus style for the text being selected. Sorry about that! Ignore that comment 😄

jughosta commented:
Okay, got 6 approvals and merging now 🙂
Thanks for the reviews!
